source/common/network/dns_impl.cc

```diff
@@ -81,32 +81,37 @@ void DnsResolverImpl::PendingResolution::onAresHostCallback(int status, int time
 
   std::list<Address::InstanceConstSharedPtr> address_list;
   if (status == ARES_SUCCESS) {
-    if (hostent->h_addrtype == AF_INET) {
-      for (int i = 0; hostent->h_addr_list[i] != nullptr; ++i) {
-        ASSERT(hostent->h_length == sizeof(in_addr));
-        sockaddr_in address;
-        memset(&address, 0, sizeof(address));
-        address.sin_family = AF_INET;
-        address.sin_port = 0;
-        address.sin_addr = *reinterpret_cast<in_addr*>(hostent->h_addr_list[i]);
-        address_list.emplace_back(new Address::Ipv4Instance(&address));
+    if (addrinfo != nullptr && addrinfo->nodes != nullptr) {
+      if (addrinfo->nodes->ai_family == AF_INET) {
+        for (const ares_addrinfo_node* ai = addrinfo->nodes; ai != nullptr; ai = ai->ai_next) {
+          sockaddr_in address;
+          memset(&address, 0, sizeof(address));
+          address.sin_family = AF_INET;
+          address.sin_port = 0;
+          address.sin_addr = reinterpret_cast<sockaddr_in*>(ai->ai_addr)->sin_addr;
+          address_list.emplace_back(new Address::Ipv4Instance(&address));
+        }
+      } else if (addrinfo->nodes->ai_family == AF_INET6) {
+        for (const ares_addrinfo_node* ai = addrinfo->nodes; ai != nullptr; ai = ai->ai_next) {
+          sockaddr_in6 address;
+          memset(&address, 0, sizeof(address));
+          address.sin6_family = AF_INET6;
+          address.sin6_port = 0;
+          address.sin6_addr = reinterpret_cast<sockaddr_in6*>(ai->ai_addr)->sin6_addr;
+          address_list.emplace_back(new Address::Ipv6Instance(address));
+        }
       }
-    } else if (hostent->h_addrtype == AF_INET6) {
-      for (int i = 0; hostent->h_addr_list[i] != nullptr; ++i) {
-        ASSERT(hostent->h_length == sizeof(in6_addr));
-        sockaddr_in6 address;
-        memset(&address, 0, sizeof(address));
-        address.sin6_family = AF_INET6;
-        address.sin6_port = 0;
-        address.sin6_addr = *reinterpret_cast<in6_addr*>(hostent->h_addr_list[i]);
-        address_list.emplace_back(new Address::Ipv6Instance(address));
-      }
-    }
-    if (!address_list.empty()) {
-      completed_ = true;
     }
   }
 
+  if (!address_list.empty()) {
+    completed_ = true;
+  }
+
+  if (addrinfo) {
```

Comment:
Maybe `if (status == ARES_SUCCESS`) {`, since on failure the contract seems to allow this to be anything... if this is the change in logic, maybe it can be just buried under the `if` above?

Reply:
I will move this one to `if (status == ARES_SUCCESS)`. I checked the code. 
https://github.com/c-ares/c-ares/blob/d7e070e7283f822b1d2787903cce3615536c5610/ares_getaddrinfo.c#L415

If the status is not success, the memory will be freed by the library.